airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/StateDeltaTracker.java

```diff
@@ -0,0 +1,171 @@
+  protected final List<byte[]> stateDeltas;
+
+  @VisibleForTesting
+  protected long remainingCapacity;
+  @VisibleForTesting
+  protected boolean capacityExceeded;
+
+  public StateDeltaTracker(final long memoryLimitBytes) {
+    this.committedStateHashes = new HashSet<>();
+    this.streamToCommittedRecords = new HashMap<>();
+    this.stateDeltas = new ArrayList<>();
+    this.remainingCapacity = memoryLimitBytes;
+    this.capacityExceeded = false;
+  }
+
+  /**
+   * Converts the given state hash and per-stream record count map into a {@code byte[]} and stores
+   * it.
+   * <p>
+   * This method leverages a synchronized block to provide thread safety between the source thread
+   * calling addState while the destination thread calls commitStateHash.
+   *
+   * @throws CapacityExceededException thrown when the memory footprint of stateDeltas exceeds
+   *         available capacity.
+   */
+  public void addState(final int stateHash, final Map<Short, Long> streamIndexToRecordCount) throws CapacityExceededException {
+    final int size = STATE_HASH_BYTES + (streamIndexToRecordCount.size() * BYTES_PER_STREAM);
+
+    if (capacityExceeded || remainingCapacity < size) {
+      capacityExceeded = true;
+      throw new CapacityExceededException("Memory capacity is exceeded for StateDeltaTracker.");
+    }
+
+    final ByteBuffer delta = ByteBuffer.allocate(size);
+
+    delta.putInt(stateHash);
+
+    for (final Map.Entry<Short, Long> entry : streamIndexToRecordCount.entrySet()) {
+      delta.putShort(entry.getKey());
+      delta.putLong(entry.getValue());
+    }
+
+    synchronized (this) {
+      stateDeltas.add(delta.array());
+      remainingCapacity -= delta.array().length;
+    }
+  }
+
+  /**
+   * Mark the given {@code stateHash} as committed.
+   * <p>
+   * This method leverages a synchronized block to provide thread safety between the source thread
+   * calling addState while the destination thread calls commitStateHash.
+   *
+   * @throws StateHashConflictException thrown when the given {@code stateHash} is already committed
+   * @throws CapacityExceededException thrown when committed counts can no longer be reliably computed
+   *         due earlier exceeded capacity from addState.
+   */
+  public void commitStateHash(final int stateHash) throws StateHashConflictException, CapacityExceededException {
+    if (capacityExceeded) {
+      throw new CapacityExceededException("Memory capacity exceeded for StateDeltaTracker, so states cannot be reliably committed");
+    }
+    if (committedStateHashes.contains(stateHash)) {
+      throw new StateHashConflictException(String.format("State hash %d was already committed, likely indicating a state hash collision", stateHash));
+    }
+
+    synchronized (this) {
+      this.committedStateHashes.add(stateHash);
+      int currStateHash;
+      do {
+        if (stateDeltas.isEmpty()) {
+          // Should never happen as long as addState always called before commitStateHash
```

Comment:
woof. glad you called this our. this assumption is not technically true. in the case where someone is replicating data using `full_refresh`. 🙄  I think this means 2 things:
1. in success cases, if everything completes successfully (i.e. source and destination contains both complete with `exit 0`, we make the assumption that all records were committed.
2. in cases where no state messages are emitted (most common case will be failures in the middle of a `full_refresh` sync), we should at least fail gracefully here. totally fine if we fall back on saying, we don't know what was committed, but we shouldn't explode.